### PR TITLE
SU(n) group

### PR DESCRIPTION
**Pull request by juli31**

### CODE REVIEW

File: sympy/liealgebras/SU.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+from sympy.matrices import zeros, ones, Matrix, randMatrix, matrices
+from sympy import mpmath as mp
```
Comment (jksuom) [inline sympy/liealgebras/SU.py]: mpmath is no longer a part of SymPy. This should probably be `import mpmath`.

File: sympy/liealgebras/SU.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+from mpmath import qr_solve
```
Comment (jksuom) [inline sympy/liealgebras/SU.py]: Are qr_solve and lu_solve from mpmath essential? There should be also SymPy versions available.

File: sympy/liealgebras/SU.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+from mpmath import lu_solve
+
+#generator of SU(n) group
```
Comment (jksuom) [inline sympy/liealgebras/SU.py]: This (or part of it) could belong the docstring. Similarly with other comments below.

File: sympy/liealgebras/SU.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+#now works only for SU(2), but since all the other functions can be
+#used for every SU(n) group, i will have to write only a new generator
+#here
+
+def SU_generator(n,str):
+    generator=[]
```
Comment (jksuom) [inline sympy/liealgebras/SU.py]: Please add a space on either side of =.

File: sympy/liealgebras/SU.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    list_of_generators=[]
+    if (n==2):
```
Comment (jksuom) [inline sympy/liealgebras/SU.py]: A space on either side of ==.

File: sympy/liealgebras/SU.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        generator.append(Matrix([[0,0+1j],[0+1j,0]]))
```
Comment (jksuom) [inline sympy/liealgebras/SU.py]: A space after each comma.

File: sympy/liealgebras/SU.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        generator.append(Matrix([[0,0-1j],[0+1j,0]]))
```
Comment (jksuom) [inline sympy/liealgebras/SU.py]: Why `0 - 1`, `0 + 1` instead of -1 and 1?

**Comment by jksuom** [comment]
Trailing whitespace can be found by running `bin/doctest SU.py`.

File: sympy/liealgebras/SU.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+from sympy import mpmath as mp
+from mpmath import qr_solve
+from mpmath import lu_solve
+
+#generator of SU(n) group
+#now works only for SU(2), but since all the other functions can be
+#used for every SU(n) group, i will have to write only a new generator
+#here
+
+def SU_generator(n,str):
```
Comment (asmeurer) [inline sympy/liealgebras/SU.py]: Call the argument something other than `str`. `str` is a builtin name in Python and it's bad to shadow those.

**Comment by jksuom** [comment]
It seems that this is concerned with the Lie algebra of SU(n), not the group itself. It is customary to use lower case letters to denote Lie algebras, hence a more suitable file name would be su.py. The name can be changed by `git mv`.

**Comment by juli31** [comment]
Thank you for your comments. Can you see if all the necessary changes were made?

**Comment by asmeurer** [comment]
Capital letters in module names also cause issues on systems with case insensitive file systems like macOS and Windows, and should be avoided anyway.

**Comment by juli31** [comment]
I think I fixed all bugs but Travis CI still reports failure of some tests which are not related to my code. Please tell me if I'm wrong.

**Comment by jksuom** [comment]
It looks like your code tries to invert a random matrix that is not invertible while running LUsolve.

BTW, there are methods in SymPy that can be used to determine if a matrix is unitary (like the elements of the group SU(n)) or antihermitian (like the elements of the Lie algebra su(n)). So there is no need for new functions for those purposes.

**Comment by czgdp1807** [comment]
Closed accidentally. @jksuom Is this PR still needed?

**Comment by jksuom** [comment]
It seems that `su_generator` will create a diagonal matrices of trace 0 and non-diagonal elementary matrices with two non-zero entries. The construction of such matrices by two loops of row_joins and col_joins looks inefficient. It would probably suffice to start with a matrix of zeros and then set the two entries at (i, j) and (j, i).

The doctring of `is_su` speaks of a matrix of the *group* but in the test a Lie algebra element is entered. The function usually fails with ValueError as most of the Lie algebra matrices are not invertible. However, the function is not needed as `.is_hermitian()` can be used instead.

The `liealgebra` module does not currently handle the actual algebra structures, but I don't think this PR would add much to it.

**Comment by czgdp1807** [comment]
Closing as the PR is inactive from quite a long time and the above comment says(according to my interpretation) doesn't add much to `master`. However if you think this can be re-opened then please let us know. Thanks for your contributions.

**Pull request closed without merging**